### PR TITLE
Internal API for reading/writing to higher protocol tables and creating higher protocol tables

### PR DESCRIPTION
**Pull request by allisonport-db**
## Overview

Adds internal APIs for testing purposes to allow creating tables of higher-protocol versions, and allow reading/writing to them. All public APIs should be unaffected and will still only support Protocol(1, 2).

### Reading/writing to higher-protocol tables
We introduce the separate concepts of standalone's supported protocol version vs. the connector's supported protocol version. To read/write to a table standalone's versions must be >= the table's protocol version _and_ the connector's versions must be >= the table's protocol version.

For now we add `connectorReaderVersion` and `connectorWriterVersion` to `DeltaLogImpl` and allow setting them internally through `DeltaLogImpl.forTable(...)` (but do NOT add it to the public API.) These are temporary and in the future will be replaced with `connectorSupportedReaderFeatures` and `connectorSupportedWriterFeatures`.

### Creating higher-protocol tables

We introduce `OptimisticTransactionImpl.upgradeProtocolVersion` which enables creating a new table with a higher protocol version than (1,2). This is not added to the public OptimisticTransaction interface--all usages will need to first cast `txn.asInstanceOf[OptimisticTransactionImpl]`

## How will these APIs be used?

We can categorize the two types of tests we will see:
1) Testing the connector protocol checks ([example](https://github.com/delta-io/connectors/pull/454/commits/973e45a5db10fee01128181f8b067a1b6b9373a6#diff-0ea3a9c91f6b9903d13d78f1fb2759d0ff3b43a6a30550547795761624c6af0dL367)). Do we correctly use the provided `supportedReaderFeatures`/`supportedWriterFeatures` to fail? (this can be failing reads/writes, failing `enableFeatureInProtocol` etc)
2) Testing feature APIs ([example](https://github.com/delta-io/connectors/pull/454/files#diff-0ea3a9c91f6b9903d13d78f1fb2759d0ff3b43a6a30550547795761624c6af0dR326)). As an example, do we verify the metadata correctly?

I've added an example of (1) to the PR but will remove before merge.

### Test utils

To facilitate tests of type (2) we add `getDeltaLogWithStandaloneAsConnector` to `TestUtils`.
- Once we switch to `connectorSupportedReaderFeatures` and `connectorSupportedWriterFeatures` we will update `getDeltaLogWithStandaloneAsConnector` to set the connector's supported feature sets to equal standalone's supported feature sets

I've also added `testException` to `TestUtils`.
  - TODO: as a followup refactor PR, update other tests to use `TestUtils`

### CODE REVIEW

File: standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionSuite.scala [hunk 1]
```diff
@@ -316,4 +318,88 @@ class OptimisticTransactionSuite extends OptimisticTransactionSuiteBase {
...
+  // prepareCommit() protocol checks
+  ///////////////////////////////////////////////////////////////////////////
+
+  test("appendOnly: metadata-protocol compatibility checks") {
+    val schema = new StructType(Array(new StructField("col1", new IntegerType(), true)))
+
+    // cannot set appendOnly=true with too low a protocol version
+    withTempDir { dir =>
+      val log = getDeltaLogWithStandaloneAsConnector(new Configuration(), dir.getCanonicalPath)
+      val txn = log.startTransactionWithInitialProtocolVersion(1, 1)
```
Comment (scottsand-db) [inline standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionSuite.scala:332]: this (1, 1) protocol version isn't a "higher" protocol version, as per the method docs of `startTransactionWithInitialProtocolVersion`.

do we need to use this API here?

File: standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala [hunk 2]
```diff
@@ -183,6 +191,25 @@ private[internal] class DeltaLogImpl private(
   // Internal Methods
   ///////////////////////////////////////////////////////////////////////////
 
+  // This is a temporary private API to create tables with a higher protocol version to allow
+  // implementing and testing higher-protocol features before we add the public APIs to upgrade
+  // table protocol versions.
+  private[standalone] def startTransactionWithInitialProtocolVersion(
```
Comment (scottsand-db) [inline standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala:197]: what will the code look like when this API is removed? what will change?

File: standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala [hunk 3]
```diff
@@ -183,6 +191,25 @@ private[internal] class DeltaLogImpl private(
   // Internal Methods
   ///////////////////////////////////////////////////////////////////////////
 
+  // This is a temporary private API to create tables with a higher protocol version to allow
```
Comment (scottsand-db) [inline standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala:194]: `internal` API

File: standalone/src/main/scala/io/delta/standalone/internal/actions/actions.scala [hunk 4]
```diff
@@ -26,12 +26,16 @@ import com.fasterxml.jackson.databind.annotation.{JsonDeserialize, JsonSerialize
...
+import io.delta.standalone.internal.DeltaConfigs
+import io.delta.standalone.internal.exception.DeltaErrors
 import io.delta.standalone.internal.util.{DataTypeParser, JsonUtils}
 
 private[internal] object Action {
   /** The maximum version of the protocol that this version of Delta Standalone understands. */
-  val maxSupportedReaderVersion = 1
-  val maxSupportedWriterVersion = 2
+  // Note: features between version (1,2) and (2,5) are in development. We bump to (2,5) now
+  // since the features are not being implemented linearly. This does not affect any public APIs.
```
Comment (scottsand-db) [inline standalone/src/main/scala/io/delta/standalone/internal/actions/actions.scala:36]: and nothing will be merged to master until all features have been completed/suppported

File: standalone/src/main/scala/io/delta/standalone/internal/actions/actions.scala [hunk 5]
```diff
@@ -61,8 +65,8 @@ private[internal] sealed trait Action {
  * fields that they do not understand.
  */
 private[internal] case class Protocol(
-    minReaderVersion: Int = Action.maxSupportedReaderVersion,
-    minWriterVersion: Int = Action.maxSupportedWriterVersion) extends Action {
```
Comment (scottsand-db) [inline standalone/src/main/scala/io/delta/standalone/internal/actions/actions.scala:65]: why this change?

File: standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionSuite.scala [hunk 6]
```diff
@@ -316,4 +318,88 @@ class OptimisticTransactionSuite extends OptimisticTransactionSuiteBase {
...
+
+  ///////////////////////////////////////////////////////////////////////////
+  // TEMPORARY (will be removed before merging)
+  ///////////////////////////////////////////////////////////////////////////
+
+  // This is to demonstrate the two types of tests we will have:
+  // (1) Testing the connector protocol checks. Do we correctly use the provided
+  //     supportedReaderFeatures/supportedWriterFeatures to fail? (this can be failing reads/writes,
+  //     failing enableFeatureInProtocol etc)
+  // (2) Testing feature APIs. As an example, do we verify the metadata correctly?
```
Comment (scottsand-db) [inline standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionSuite.scala:373]: can't each of these go into separate test suites? would that be a clean organization?

File: standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala (hunk 2 above)
Comment (scottsand-db) [inline standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala:197]: what will the public API be and why can't we implement it now?

File: standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionSuite.scala (hunk 1 above)
Comment (allisonport-db) [reply standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionSuite.scala:332]: If we don't do this, we create a table with Protocol(1, 2) by default currently. Updating our default protocol will be a part of the later protocol API updates.

Comment (allisonport-db) [reply standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionSuite.scala:332]: I can add the docs to be more specific. As in, it's a method to specify the initial protocol version for a newly created table.

File: standalone/src/main/scala/io/delta/standalone/internal/actions/actions.scala (hunk 5 above)
Comment (allisonport-db) [reply standalone/src/main/scala/io/delta/standalone/internal/actions/actions.scala:65]: All the usages of `Protocol()` in the code were now using `Protocol(2, 5)` when they expected `Protocol(1,2)`. There's also no real reason to justify this default being a specific version now that we are supporting more than a single protocol and it makes the code more confusing / less-self-documenting. I.e. seeing `Protocol()` in the code is confusing and ambiguous

File: standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionSuite.scala (hunk 6 above)
Comment (allisonport-db) [reply standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionSuite.scala:373]: I think (1) will likely go into a separate suite. I was thinking (2) will likely be more broken down by feature and go in feature-specific suites. `appendOnly` is a weird case where we already support it and are now adding the check

Comment (allisonport-db) [reply standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionSuite.scala:373]: If you think the protocol-metadata compatibility checks should go all in the same suite for all the featues--that could be a valid organization and I could move the appendOnly test there

File: standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala (hunk 2 above)
Comment (allisonport-db) [reply standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala:197]: Mentioned in the description above
```
val txn = log.startTransaction()
txn.upgradeProtocolVersion(1, 3)
```
Although I'm now wondering whether it should be `setProtocolVersion` instead to account for the first version of the table

Comment (allisonport-db) [reply standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala:197]: somewhat answered above. we decided to implement the protocol-specific public APIs later. This will involve additional checks (like making sure connectors don't block themselves out) & is just additional work compared to the "lightweight" internal API we provide here

File: standalone/src/main/scala/io/delta/standalone/internal/actions/actions.scala (hunk 4 above)
Comment (allisonport-db) [reply standalone/src/main/scala/io/delta/standalone/internal/actions/actions.scala:36]: Is not changing any public APIs not sufficient?

Comment (scottsand-db) [reply standalone/src/main/scala/io/delta/standalone/internal/actions/actions.scala]: it may or may not be, but making it clear it won't be merged until finished removes that ambiguity.

e.g. what if it doesn't change any public API, but changes some behind-the-scenes internal implementation?

File: standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala (hunk 2 above)
Comment (scottsand-db) [reply standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala:197]: if the future public API will be `txn.upgradeProtocolVersion(1, 3)`, why don't we just implement it now (public) but with your mock/lightweight impl?

that means all our tests can use the public API, and we can improve it gradually.

Comment (allisonport-db) [reply standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala:197]: I think we don't want to add public APIs without tests. I would prefer the extra annoyance of refactoring later

We could also leave `startTransactionWithInitialProtocolVersion` as an internal testing method in `DeltaLogImpl` and just update it to do the above snippet.

File: standalone/src/main/scala/io/delta/standalone/internal/actions/actions.scala (hunk 4 above)
Comment (allisonport-db) [reply standalone/src/main/scala/io/delta/standalone/internal/actions/actions.scala]: Err then if that affects a public API it would be changing the public API? I've updated it either way

File: standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionLegacySuite.scala [hunk 7]
```diff
@@ -250,25 +251,27 @@ class OptimisticTransactionLegacySuite extends FunSuite {
     withTempDir { dir =>
       val log = DeltaLog.forTable(new Configuration(), dir.getCanonicalPath)
 
-      Seq(Protocol(1, 3), Protocol(1, 1), Protocol(2, 2)).foreach { protocol =>
```
Comment (allisonport-db) [inline standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionLegacySuite.scala:253]: Here and below: Should we allow tables to be created with Protocol(1,1)? In the current version we have blocked this.

I understand (1, 2) being the default, but I'm unclear on whether we should explicitly block (1,1) (we allow it in delta-spark)

File: standalone/src/main/scala/io/delta/standalone/internal/OptimisticTransactionImpl.scala [hunk 8]
```diff
@@ -289,8 +319,6 @@ private[internal] class OptimisticTransactionImpl(
       case _ => // nothing
     }
 
-    deltaLog.assertProtocolWrite(snapshot.protocolScala)
```
Comment (allisonport-db) [inline standalone/src/main/scala/io/delta/standalone/internal/OptimisticTransactionImpl.scala:292]: we do this when starting the transaction instead. we perform the exact same check then anyways we may as well fail earlier

File: standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala (hunk 2 above)
Comment (vkorukanti) [reply standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala:197]: One qn @allisonport-db : when you say later, it will be before we merge the feature branch to master right?

Comment (allisonport-db) [reply standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala:197]: later = before we do a release with any of this code merged

Comment (scottsand-db) [reply standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala:197]: `setProtocolVersion` is better

File: standalone/src/main/scala/io/delta/standalone/internal/OptimisticTransactionImpl.scala [hunk 9]
```diff
@@ -224,6 +225,25 @@ private[internal] class OptimisticTransactionImpl(
     snapshot.transactions.getOrElse(id, -1L)
   }
 
+  ///////////////////////////////////////////////////////////////////////////
+  // Internal Test-Only Methods
+  ///////////////////////////////////////////////////////////////////////////
+
+  // This is a lightweight version of the eventual public upgrade protocol by version API. When this
+  // is formally added it will be added to OptimisticTransaction and additional checks will be
```
Comment (scottsand-db) [inline standalone/src/main/scala/io/delta/standalone/internal/OptimisticTransactionImpl.scala:233]: what sort of additional checks? e.g. why doesn't this internal API check that the connector is upgrading to a new readerVersion, writerVersion that the connector itself can support?

Overall, it's hard to know if this method is purposefully or accidentally missing a validation/check

File: standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala [hunk 10]
```diff
@@ -183,6 +191,25 @@ private[internal] class DeltaLogImpl private(
...
+      readerVersion: Int, writerVersion: Int): OptimisticTransaction = {
+
+    update()
+    assertProtocolWrite(snapshot.protocolScala)
+    if (snapshot.version >= 0) {
+      // To keep the scope small, we only support setting the initial protocol version and not
+      // upgrading the version for existing tables
+      throw new UnsupportedOperationException(
+        "Setting the initial protocol is only supported when committing table version 0")
+    }
```
Comment (scottsand-db) [inline standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala:207]: why this? the future public API won't have this, right? so why are we doing this now?

File: standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionLegacySuite.scala (hunk 7 above)
Comment (scottsand-db) [reply standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionLegacySuite.scala:253]: > In the current version we have blocked this.

then let's keep blocking this. we can make an issue and add it to the project plan to implement this later. will this decision make this PR smaller?

File: standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionSuite.scala [hunk 11]
```diff
@@ -316,4 +318,88 @@ class OptimisticTransactionSuite extends OptimisticTransactionSuiteBase {
...
+
+  // The test below is an example of (1). The test above is an example of (2)
+
+  test("example of type (1) test") {
```
Comment (scottsand-db) [inline standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionSuite.scala:377]: IMO the test name should be more specific, and a comment above it can describe that it is an example of a type 1 test

File: standalone/src/test/scala/io/delta/standalone/internal/util/TestUtils.scala [hunk 12]
```diff
@@ -47,4 +50,16 @@ object TestUtils {
...
 
+  def testException[T <: Throwable](f: => Any, messageContains: String)
+    (implicit manifest: Manifest[T]): Unit = {
+    val e = intercept[T]{
+      f
+    }.getMessage
+    assert(e.contains(messageContains))
+  }
+
+  def getDeltaLogWithStandaloneAsConnector(conf: Configuration, path: String) : DeltaLogImpl = {
```
Comment (scottsand-db) [inline standalone/src/test/scala/io/delta/standalone/internal/util/TestUtils.scala:61]: is this temporary?

also, method docs since this is an interesting / bold function name

File: standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionLegacySuite.scala (hunk 7 above)
Comment (allisonport-db) [reply standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionLegacySuite.scala:253]: It won't really make it smaller but it will reduce the scope. This can be a part of the general protocol changes we will make.

It will also make the `appendOnly` tests obsolete though, since we won't be creating any tables with Protocol(1,1)

File: standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionSuite.scala (hunk 11 above)
Comment (allisonport-db) [reply standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionSuite.scala:377]: I was planning on removing this before merging the PR was just for demonstration purposes

File: standalone/src/test/scala/io/delta/standalone/internal/util/TestUtils.scala (hunk 12 above)
Comment (allisonport-db) [reply standalone/src/test/scala/io/delta/standalone/internal/util/TestUtils.scala:61]: I think this makes sense to keep as a test util going forward. Once we switch from protocol version --> feature lists, this will set the connector's supported feature sets to equal standalone's supported feature sets. Ack on the method docs

File: standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala (hunk 2 above)
Comment (vkorukanti) [reply standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala:197]: Once `txn.upgradeProtocolVersion(1, 3)` is ready, is this just to set the transaction on new table or can it be used to upgrade the protocol on existing table?

File: standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala [hunk 13]
```diff
@@ -253,13 +298,27 @@ private[standalone] object DeltaLogImpl {
     apply(hadoopConf, new Path(dataPath, "_delta_log"), clock)
   }
 
+  def forTable(hadoopConf: Configuration, dataPath: String,
```
Comment (vkorukanti) [inline standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala:301]: Is this a public API? may need docs if it is.

File: standalone/src/main/scala/io/delta/standalone/internal/actions/actions.scala [hunk 14]
```diff
@@ -79,6 +84,32 @@ private[internal] object Protocol {
...
+
+    // Append-only
+    if (DeltaConfigs.IS_APPEND_ONLY.fromMetadata(metadata) && protocol.minWriterVersion < 2) {
+      // todo: when we add the feature enums we can use their string here (and all below)
+      throw DeltaErrors.insufficientWriterVersion(protocol, 2, "appendOnly")
+    }
+
+    // Check constraints
+
+    // Generated columns
```
Comment (vkorukanti) [inline standalone/src/main/scala/io/delta/standalone/internal/actions/actions.scala:103]: are all the features here can be verified using the properties in Metadata?

File: standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala (hunk 10 above)
Comment (allisonport-db) [reply standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala:207]: so we don't have to implement upgrade checks + the tests for it. I can remove the check but initial protocol is sufficient for our tests (we may as well allow only what we need, not more)

File: standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala (hunk 13 above)
Comment (allisonport-db) [reply standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala:301]: not public

File: standalone/src/main/scala/io/delta/standalone/internal/actions/actions.scala (hunk 14 above)
Comment (allisonport-db) [reply standalone/src/main/scala/io/delta/standalone/internal/actions/actions.scala:103]: yes, we'll basically do the same thing as in `Protocol.requiredMinimumProtocol` in https://github.com/delta-io/delta (and add each check as we add the feature)

File: standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala (hunk 2 above)
Comment (allisonport-db) [reply standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala]: removed this in favor of just directly using `upgradeProtocolVersion`

File: standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala (hunk 13 above)
Comment (scottsand-db) [reply standalone/src/main/scala/io/delta/standalone/internal/DeltaLogImpl.scala:282]: @vkorukanti this API is in `DeltaLogImpl.scala` and not `DeltaLog.java`. Internal tests will create `DeltaLogImpl`s.

the future public API will use features, not integer versions.

File: standalone/src/main/scala/io/delta/standalone/internal/OptimisticTransactionImpl.scala [hunk 15]
```diff
@@ -277,8 +293,19 @@ private[internal] class OptimisticTransactionImpl(
 
     val protocolOpt = finalActions.collectFirst{ case p: Protocol => p }
     if (protocolOpt.isDefined) {
```
Comment (scottsand-db) [inline standalone/src/main/scala/io/delta/standalone/internal/OptimisticTransactionImpl.scala:295]: nit: since we are do so many `protocolOpt.get` we may as well do
`protocolOpt.foreach { protocol => ... }`

File: standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionLegacySuite.scala [hunk 16]
```diff
@@ -251,24 +252,26 @@ class OptimisticTransactionLegacySuite extends FunSuite {
       val log = DeltaLog.forTable(new Configuration(), dir.getCanonicalPath)
 
       Seq(Protocol(1, 3), Protocol(1, 1), Protocol(2, 2)).foreach { protocol =>
-        val e = intercept[AssertionError] {
+        val e = intercept[Throwable] {
```
Comment (scottsand-db) [inline standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionLegacySuite.scala:255]: can we catch anything more specific?

File: standalone/src/test/scala/io/delta/standalone/internal/util/TestUtils.scala (hunk 12 above)
Comment (scottsand-db) [reply standalone/src/test/scala/io/delta/standalone/internal/util/TestUtils.scala:65]: idea: `getDeltaLogWithMaxFeatureSupport` ? im easy.

**Review by scottsand-db — APPROVED** [review]
LGTM with one nit and one method rename suggestion.

File: standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionLegacySuite.scala (hunk 16 above)
Comment (allisonport-db) [reply standalone/src/test/scala/io/delta/standalone/internal/OptimisticTransactionLegacySuite.scala:255]: Not really since we're explicitly blocking the writerVersion=1 with an assert. But we see `InvalidProtocolVersionException` for protocols > Protocol(1,2)

Way around this would be to test Protocol(1,3) and Protocol(2,2) separately from Protocol(1,1). I'm good with either here since once we later allow writerVersion=1 we can update this to be more specific
